components/eamxx/src/share/util/eamxx_time_interpolation.cpp

```diff
@@ -0,0 +1,340 @@
+{
+  m_time0 = ts_in;
+  m_time1 = ts_in;
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which will initialize field data given an input field.
+ * Input:
+ *   field_in - A field with a name matching one of the fields in the interpolator.  Data will be
+ *              shifted and copied.
+ */
+void TimeInterpolation::initialize_data_from_field(const Field& field_in)
+{
+  const auto name = field_in.name();
+  auto field0 = m_fm_time0->get_field(name);
+  auto field1 = m_fm_time1->get_field(name);
+  field0.deep_copy(field_in);
+  field1.deep_copy(field_in);
+  auto ts = field_in.get_header().get_tracking().get_time_stamp();
+  m_time0 = ts;
+  m_time1 = ts;
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which will initialize the full set of fields given a list of files pointing to where
+ * the data can be found.
+ * Input:
+ *   list_of_files: A vector of strings representing all the files where interpolation data can be
+ *                  gathered.
+ */
+void TimeInterpolation::initialize_data_from_files()
+{
+  // Initialize the AtmosphereInput object that will be used to gather data
+  ekat::ParameterList input_params;
+  input_params.set("Field Names",m_field_names);
+  input_params.set("Filename",m_triplet_iterator->filename);
+  m_file_data_atm_input = AtmosphereInput(input_params,m_fm_time1);
+  // Read first snap of data and shift to time0
+  read_data();
+  shift_data();
+  update_timestamp(m_triplet_iterator->timestamp);
+  // Advance the iterator and read the next set of data for time1
+  ++m_triplet_iterator;
+  read_data();
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which will update the timestamps by shifting time1 to time0 and setting time1.
+ * Input:
+ *   ts_in - A timestamp for the most recent timestamp of the interpolation data.
+ *
+ * It is assumed that any updated time is meant to replace time1 and that the time1
+ * should be shifted to time0
+ */
+void TimeInterpolation::update_timestamp(const TimeStamp& ts_in)
+{
+  m_time0 = m_time1;
+  m_time1 = ts_in;
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which will update field data given an input field.  Useful for time interpolation
+ * related to the EAMxx state.
+ * Input:
+ *   field_in - A field with a name matching one of the fields in the interpolator.  Data will be
+ *              shifted and copied.
+ *
+ * It is assumed that any updated data is meant to replace the data at time1 and that the time1
+ * data should be shifted to time0
+ */
+void TimeInterpolation::update_data_from_field(const Field& field_in)
+{
+  const auto name = field_in.name();
+  auto field0 = m_fm_time0->get_field(name);
+  auto field1 = m_fm_time1->get_field(name);
+  field0.deep_copy(field1);
```

Comment:
Another place where std::swap should be used, to avoid the f1->f0 deep copy (we'll still need the deep copy f_in->f1.